src/lava_callback.py

```diff
@@ -187,9 +202,194 @@ async def callback(node_id: str, request: Request):
+
+
+def validate_permissions(jwtoken, permission):
+    if not jwtoken:
+        return False
+    try:
+        decoded = decode_jwt(jwtoken)
+    except Exception as e:
+        logging.error(f'Error decoding JWT: {e}')
+        return False
+    if not decoded:
+        logging.error('Invalid JWT')
+        return False
+    permissions = decoded.get('permissions')
+    if not permissions:
+        logging.error('No permissions in JWT')
+        return False
+    if permission not in permissions:
+        logging.error(f'Permission {permission} not in JWT')
+        return False
+    return decoded
+
+
+def find_parent_kind(node, api_helper, kind):
+    parent_id = node.get('parent')
+    if not parent_id:
+        return None
+    parent_node = api_helper.api.node.get(parent_id)
+    if not parent_node:
+        return None
+    if parent_node.get('kind') == kind:
+        return parent_node
+    return find_parent_kind(parent_node, api_helper, kind)
+
+
+@app.post('/api/jobretry')
+async def jobretry(data: JobRetry, request: Request,
+                   Authorization: str = Header(None)):
+    '''
+    API call to assist in regression bisecting by retrying a specific job
+    retrieved from test results.
+    '''
+    # Validate JWT token from Authorization header
+    jwtoken = Authorization
+    decoded = validate_permissions(jwtoken, 'testretry')
+    if not decoded:
+        return 'Unauthorized', 401
+
+    email = decoded.get('email')
+    logging.info(f"User {email} is retrying job {data.nodeid}")
+    api_config_name = SETTINGS.get('DEFAULT', {}).get('api_config')
+    if not api_config_name:
+        return 'No default API name set', 500
+    api_token = os.getenv('KCI_API_TOKEN')
+    api_helper = _get_api_helper(api_config_name, api_token)
+    node = api_helper.api.node.get(data.nodeid)
+    if not node:
+        return 'Node not found', 404
+    if node['kind'] != 'job':
+        return 'Node is not a job', 400
+
+    knode = find_parent_kind(node, api_helper, 'kbuild')
+    if not knode:
+        return 'Kernel build not found', 404
+
+    jobfilter = [knode['name'], node['name']]
+    knode['jobfilter'] = jobfilter
+    knode['op'] = 'updated'
+    knode['data'].pop('artifacts', None)
+    # state - done, result - pass
+    if knode.get('state') != 'done':
+        return 'Kernel build is not done', 400
+    if knode.get('result') != 'pass':
+        return 'Kernel build result is not pass', 400
+    # remove created, updated, timeout, owner, submitter, usergroups
+    knode.pop('created', None)
+    knode.pop('updated', None)
+    knode.pop('timeout', None)
+    knode.pop('owner', None)
+    knode.pop('submitter', None)
+    knode.pop('usergroups', None)
+
+    evnode = {'data': knode}
+    # Now we can submit custom kbuild node to the API(pub/sub)
+    api_helper.api.send_event('node', evnode)
+    logging.info(f"Job retry for node {data.nodeid} submitted")
+    return 'OK', 200
+
+
+@app.post('/api/checkout')
```

Comment:
This additional API endpoint benefits only from being run with the pipeline service privileges - other than that it could be a `kci` tool subcommand.

I get its use case but I don't believe it solves the real issue here - Node permissions and access control, they will have to be addressed sooner or later.